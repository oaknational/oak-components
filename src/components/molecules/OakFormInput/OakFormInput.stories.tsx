import { Meta, StoryObj } from "@storybook/react";

import { OakFormInput } from "./OakFormInput";

import { sizeArgTypes } from "@/storybook-helpers/sizeStyleHelpers";

const meta: Meta<typeof OakFormInput> = {
  component: OakFormInput,
  tags: ["autodocs"],
  title: "components/molecules/OakFormInput",
  argTypes: {
    wrapperWidth: sizeArgTypes["$width"],
    wrapperMaxWidth: sizeArgTypes["$width"],
  },
  parameters: {
    controls: {
      include: [
        "placeholder",
        "value",
        "invalid",
        "disabled",
        "wrapperWidth",
        "wrapperMaxWidth",
      ],
    },
  },
  args: {
    placeholder: "Placeholder text",
  },
};
export default meta;

type Story = StoryObj<typeof OakFormInput>;

export const Default: Story = {};

export const Disabled: Story = {
  args: {
    value: "A fine text value",
    disabled: true,
  },
};

export const Invalid: Story = {
  args: {
    value: "A wrong text value",
    invalid: true,
  },
};
