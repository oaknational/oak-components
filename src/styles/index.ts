export * from "@/styles/theme/color";
