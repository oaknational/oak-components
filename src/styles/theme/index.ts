export { oakColorTokens, oakUiRoleTokens } from "./color";

export type {
  OakColorToken,
  OakCombinedColorToken,
  OakUiRoleToken,
  OakColorFilterToken,
} from "./color";

export {
  oakAllSpacingTokens,
  oakInnerPaddingTokens,
  oakSpaceBetweenTokens,
} from "./spacing";

export type {
  OakAllSpacingToken,
  OakCombinedSpacingToken,
  OakInnerPaddingToken,
  OakSpaceBetweenToken,
} from "./spacing";

export { oakDropShadowTokens } from "./dropShadow";
export type { OakDropShadowToken } from "./dropShadow";

export { oakOpacityTokens } from "./opacity";
export type { OakOpacityToken } from "./opacity";

export { oakTransitionTokens } from "./transitions";
export type { OakTransitionToken } from "./transitions";

export { oakBorderRadiusTokens, oakBorderWidthTokens } from "./borders";
export type { OakBorderRadiusToken, OakBorderWidthToken } from "./borders";

export type { OakTheme } from "./theme";

export { oakFontTokens, oakFontSizeTokens } from "./typography";
export type { OakFontToken, OakFontSizeToken } from "./typography";

export { oakZIndexTokens } from "./zIndex";

export { oakDefaultTheme } from "./default.theme";
