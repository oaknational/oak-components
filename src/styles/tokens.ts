export const oakAllSpacingPx = {
  "all-spacing-1": 4,
  "all-spacing-2": 8,
  "all-spacing-3": 12,
  "all-spacing-4": 16,
  "all-spacing-5": 20,
  "all-spacing-6": 24,
  "all-spacing-7": 32,
  "all-spacing-8": 40,
  "all-spacing-9": 48,
  "all-spacing-10": 56,
  "all-spacing-11": 64,
  "all-spacing-12": 72,
  "all-spacing-13": 80,
  "all-spacing-14": 92,
  "all-spacing-15": 100,
  "all-spacing-16": 120,
};

export type OakAllSpacing = keyof typeof oakAllSpacingPx;

export const oakInnerPaddingAllSpacing = {
  "inner-padding-xs": "all-spacing-2",
  "inner-padding-s": "all-spacing-3",
  "inner-padding-m": "all-spacing-4",
  "inner-padding-l": "all-spacing-5",
  "inner-padding-xl": "all-spacing-6",
};

export type OakInnerPadding = keyof typeof oakInnerPaddingAllSpacing;

export const oakSpaceBetweenAllSpacing = {
  "space-between-ssx": "all-spacing-2",
  "space-between-xs": "all-spacing-3",
  "space-between-s": "all-spacing-4",
  "space-between-m": "all-spacing-6",
  "space-between-m2": "all-spacing-7",
  "space-between-l": "all-spacing-9",
  "space-between-xl": "all-spacing-10",
  "space-between-xxl": "all-spacing-12",
  "space-between-xxxl": "all-spacing-13",
};

export type OakSpaceBetween = keyof typeof oakSpaceBetweenAllSpacing;

export const oakAllColorsHex = {
  white: "#ffffff",
  grey20: "#f2f2f2",
  grey30: "#e4e4e4",
  grey40: "#cacaca",
  grey50: "#808080",
  grey60: "#575757",
  grey70: "#2d2d2d",
  black: "#222222",
  oakGreen: "#287c34",
  mint: "bef2bd",
  mint30: "ebfbeb",
  mint50: "dff9de",
  aqua: "#b0e2de",
  aqua30: "#e7f6f5",
  aqua50: "#cee7e5",
  lavender: "#a0b6f2",
  lavender30: "#e3e9fb",
  lavender50: "#c6d1ef",
  pink: "#deb7d5",
  pink30: "#f5e9f2",
  pink50: "#e5d1e0",
  lemon: "#ffe555",
  lemon30: "#fff7cc",
  lemon50: "#f6e8a0",
  amber: "#ff934e",
  amber30: "#ffdfca",
  amber50: "#ffc8a6",
  red: "#dd0035",
  navy: "#0d24c4",
  blue: "#374cf1",
  magenta: "#d02aa7",
  purple: "#845ad9",
  teal: "#037b7d",
};

export type OakColor = keyof typeof oakAllColorsHex;

export const oakColorTokens = [
  "text-primary",
  "text-subdued",
  "text-error",
  "text-disabled",
  "text-link-active",
  "text-link-hover",
  "text-link-visited",
  "text-inverted",
  "text-success",
  "text-warning",
  "bg-primary",
  "bg-neutral",
  "bg-neutral-50",
  "bg-neutral-30",
  "bg-btn-primary",
  "bg-btn-primary-hover",
  "bg-btn-primary-disabled",
  "bg-btn-secondary",
  "bg-btn-secondary-hover",
  "bg-btn-secondary-disabled",
  "bg-icon",
  "bg-icon-hover",
  "bg-decorative1â€“main",
  "bg-decorative1-subdued",
  "bg-decorative1-very subdued",
  "bg-decorative2-main",
  "bg-decorative2-subdued",
  "bg-decorative2-very-subdued",
  "bg-decorative3",
  "bg-decorative3-50",
  "bg-decorative3-30",
  "bg-decorative4",
  "bg-decorative4-50",
  "bg-decorative4-30",
  "bg-decorative5",
  "bg-decorative5-50",
  "bg-decorative5-30",
  "bg-decorative6",
  "bg-decorative6-50",
  "bg-decorative6-30",
  "border-primary",
  "border-inverted",
  "border-neutral",
  "border-decorative1",
  "border-decorative2",
  "border-decorative3",
  "border-decorative4",
  "border-decorative5",
  "border-decorative6",
] as const;

export type OakColorToken = (typeof oakColorTokens)[number];
