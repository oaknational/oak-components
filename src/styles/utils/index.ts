export * from "./colorStyle";
export * from "./positionStyle";
export * from "./sizeStyle";
export * from "./spacingStyle";
export * from "./flexStyle";
