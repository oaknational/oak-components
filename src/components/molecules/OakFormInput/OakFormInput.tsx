import React, { ChangeEventHandler } from "react";
import styled, { css } from "styled-components";

import { parseColor } from "@/styles/helpers/parseColor";
import {
  InternalTextInput,
  InternalTextInputProps,
} from "@/components/atoms/InternalTextInput";
import { OakFlex } from "@/components/atoms";
import { OakCombinedColorToken } from "@/styles";
import { SizeStyleProps } from "@/styles/utils/sizeStyle";

type StyledTextInputWrapperProps = {
  $disabled: boolean;
  $color: OakCombinedColorToken;
  $background: OakCombinedColorToken;
  $borderColor: OakCombinedColorToken;
  $hoverBackground: OakCombinedColorToken;
  $hoverBorderColor: OakCombinedColorToken;
  $focusBorderColor: OakCombinedColorToken;
  $focusBackgroundColor: OakCombinedColorToken;
  $disabledColor: OakCombinedColorToken;
  $disabledBorderColor: OakCombinedColorToken;
  $disabledBackgroundColor: OakCombinedColorToken;
};

export type OakFormInputProps = {
  /**
   * Disables user input and updates the appearance accordingly.
   */
  disabled?: boolean;

  /**
   * Sets the value. Use this in controlled components;
   */
  value?: string;
  /**
   * Sets the initial value. Use this for an uncontrolled component;
   */
  defaultValue?: string;
  /**
   * Used to target the input element in tests.
   */
  "data-testid"?: string;
  /**
   * Callback function that is called when the input value changes.
   */
  onChange?: ChangeEventHandler<HTMLInputElement>;
  /**
   * Alters the appearance of the input field to indicate when the input is invalid.
   */
  invalid?: boolean;
  /**
   * Give the field a highlight to draw attention to it
   */
  invalidBorderColor?: OakCombinedColorToken;
  color?: OakCombinedColorToken;
  background?: OakCombinedColorToken;
  borderColor?: OakCombinedColorToken;
  hoverBackground?: OakCombinedColorToken;
  hoverBorderColor?: OakCombinedColorToken;
  focusBorderColor?: OakCombinedColorToken;
  focusBackgroundColor?: OakCombinedColorToken;
  disabledColor?: OakCombinedColorToken;
  disabledBorderColor?: OakCombinedColorToken;
  disabledBackgroundColor?: OakCombinedColorToken;
  /**
   * The width of the surrounding div - the input will fill this with padding.
   */
  wrapperWidth?: SizeStyleProps["$width"];
  wrapperMaxWidth?: SizeStyleProps["$maxWidth"];
} & InternalTextInputProps;

const StyledTextInputWrapper = styled(OakFlex)<StyledTextInputWrapperProps>`
  &:hover {
    cursor: text;
  }

  background: ${(props) => parseColor(props.$background)};

  &:focus-within {
    border-color: ${(props) => parseColor(props.$focusBorderColor)};
    background: ${(props) => parseColor(props.$focusBackgroundColor)};
  }

  ${(props) => css`
    @media (hover: hover) {
      &:hover:not(:focus-within) {
        background: ${parseColor(props.$hoverBackground)};
        border-color: ${parseColor(props.$hoverBorderColor)};
      }
    }
  `}

  ${(props) =>
    props.$disabled &&
    css`
      background: ${parseColor(props.$disabledBackgroundColor)};
      border-color: ${parseColor(props.$disabledBorderColor)};
      color: ${parseColor(props.$disabledColor)};
      &:hover {
        cursor: not-allowed;
      }
    `}
`;

/**
 * Default input which can be extended to create specialised inputs.
 */
export const OakFormInput = ({
  type = "text",
  borderColor = "border-neutral-lighter",
  focusBorderColor = "border-primary",
  focusBackgroundColor = "bg-primary",
  background = "bg-primary",
  hoverBackground = "bg-neutral",
  hoverBorderColor = "border-neutral",
  disabledBackgroundColor = "bg-neutral",
  color = "text-primary",
  disabledBorderColor = "border-neutral-lighter",
  disabledColor = "text-disabled",
  invalid = false,
  invalidBorderColor = "border-error",
  wrapperWidth,
  wrapperMaxWidth,
  ...props
}: OakFormInputProps) => {
  return (
    <StyledTextInputWrapper
      $height="fit-content"
      $width={wrapperWidth}
      $maxWidth={wrapperMaxWidth}
      $borderRadius="border-radius-m"
      $ba="border-solid-m"
      $alignItems="center"
      $position="relative"
      $pa="inner-padding-xs"
      $font="body-2"
      $color={color}
      $borderColor={invalid ? invalidBorderColor : borderColor}
      $background={background}
      $hoverBackground={hoverBackground}
      $hoverBorderColor={hoverBorderColor}
      $disabledColor={disabledColor}
      $disabledBackgroundColor={disabledBackgroundColor}
      $disabledBorderColor={disabledBorderColor}
      $focusBorderColor={focusBorderColor}
      $focusBackgroundColor={focusBackgroundColor}
      $disabled={!!props.disabled}
      onClick={(event) => {
        event.currentTarget.querySelector("input")?.focus();
      }}
    >
      <InternalTextInput
        type={type}
        $width={"100%"}
        aria-invalid={invalid}
        {...props}
      />
    </StyledTextInputWrapper>
  );
};
