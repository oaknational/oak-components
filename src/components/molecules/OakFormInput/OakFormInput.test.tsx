import React from "react";
import { create } from "react-test-renderer";
import "@testing-library/jest-dom";
import { ThemeProvider } from "styled-components";

import { OakFormInput } from "./OakFormInput";

import renderWithTheme from "@/test-helpers/renderWithTheme";
import { oakDefaultTheme } from "@/styles";

describe("OakFormInput", () => {
  it("should render correctly", () => {
    const { getByPlaceholderText } = renderWithTheme(
      <OakFormInput value="" placeholder="Test input" />,
    );
    expect(getByPlaceholderText("Test input")).toBeInTheDocument();
  });

  it("should match snapshot", () => {
    const tree = create(
      <ThemeProvider theme={oakDefaultTheme}>
        <OakFormInput value="Sample text" placeholder="Test input" />
      </ThemeProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should handle disabled state", () => {
    const { getByPlaceholderText } = renderWithTheme(
      <OakFormInput value="" placeholder="Disabled input" disabled />,
    );
    const input = getByPlaceholderText("Disabled input");
    expect(input).toBeDisabled();
  });

  it("should handle invalid state", () => {
    const { getByPlaceholderText } = renderWithTheme(
      <OakFormInput value="" placeholder="Invalid input" invalid />,
    );
    const input = getByPlaceholderText("Invalid input");
    expect(input).toHaveAttribute("aria-invalid", "true");
  });
});
