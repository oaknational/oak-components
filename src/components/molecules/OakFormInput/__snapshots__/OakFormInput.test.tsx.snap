// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakFormInput should match snapshot 1`] = `
.c3 {
  -webkit-appearance: none !important;
  -moz-appearance: none !important;
  appearance: none !important;
  border: 0;
  border-radius: 0;
  border-color: transparent;
  box-shadow: none;
  font-family: inherit;
  background: transparent;
  outline: none;
  color: inherit;
  width: 100%;
}

.c3::-webkit-input-placeholder {
  color: #575757;
}

.c3::-moz-placeholder {
  color: #575757;
}

.c3:-ms-input-placeholder {
  color: #575757;
}

.c3::placeholder {
  color: #575757;
}

.c3::-webkit-search-decoration,
.c3::-webkit-search-cancel-button,
.c3::-webkit-search-results-button,
.c3::-webkit-search-results-decoration {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c3:disabled {
  cursor: not-allowed;
}

.c0 {
  position: relative;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  padding: 0.5rem;
  color: #222222;
  background: #ffffff;
  border: 0.125rem solid;
  border-color: #cacaca;
  border-radius: 0.375rem;
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 300;
  font-size: 1rem;
  line-height: 1.5rem;
  -webkit-letter-spacing: -0.005rem;
  -moz-letter-spacing: -0.005rem;
  -ms-letter-spacing: -0.005rem;
  letter-spacing: -0.005rem;
}

.c0:hover {
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  background: #ffffff;
}

.c2:hover {
  cursor: text;
}

.c2:focus-within {
  border-color: #222222;
  background: #ffffff;
}

@media (max-width:750px) {
  .c3 {
    font-size: 16px;
  }
}

@media (hover:hover) {
  .c2:hover:not(:focus-within) {
    background: #f2f2f2;
    border-color: #808080;
  }
}

<div
  className="c0 c1 c2"
  onClick={[Function]}
>
  <input
    aria-invalid={false}
    className="c3"
    onFocus={[Function]}
    placeholder="Test input"
    type="text"
    value="Sample text"
  />
</div>
`;
